_posts/2022-01-13-spice-up-your-javascript-with-curry.md

```diff
@@ -0,0 +1,314 @@
+
+```javascript
+// using arrow function expression style
+// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions
+const add = (a, b) => a + b;
+
+// using function expression style
+// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/function
+function add(a, b) {
+  return a + b;
+}
+```
+
+Manually currying this function would look like this:
+
+```javascript
+// using fat arrow
+const add = (a) => (b) => a + b;
+
+// using function keyword
+function add(a) {
+  return function (b) {
+    return a + b;
+  };
+}
+
+// As an aside, this is the last example where I use the function keyword.
+```
+
+In addition to currying there are also [_partially applied_](https://en.wikipedia.org/wiki/Partial_application) functions. The significant difference between a curried and partially applied function is that curried functions always take parameters one at a time (unary) while partially applied functions can be called multiple parameters at a time (variadic).
+
+```javascript
+// function for summing 3 numbers
+const add = (a, b, c) => a + b + c;
+
+// curried function
+const curriedAdd = (a) => (b) => (c) => add(a, b, c);
+const curriedAdd2 = curriedAdd(2);
+const curriedAdd5Add2 = curriedAdd2(5);
+curriedAdd5Add2(10) === 17; // true
+
+//partially applied function
+const partialAdd2 = add.bind(null, 2);
+partialAdd2(5, 10) === 17; // true
+isNaN(partialAdd2(5)); // true
+```
+
+However, in this form, both currying and partial application comes with caveats. You cannot call a curried function written in this form with multiple parameters at a time, since a normal function that is curried always returns a _unary_ function. This is a function that has an _arity_ of 1 meaning it only takes one argument.
+
+Partial application, however, does not protect you from calling the function with incomplete arguments before evaluating the function, which is what happened on the last row of the previous example. With currying the function is invoked lazily and it runs only after getting all the parameters.
+
+Currying with a language that does not natively understand currying is a chore. In ML-type languages (f.ex. F#) all functions are automatically curried, and the syntax supports that with proper developer ergonomics.
+
+```fsharp
+// F# example of curried function
+let add a b c = a + b + c
+
+let add2 = add 2
+let add5add2 = add2 5
+(add5add2 10) = 17 // true
+
+// However with F# we can skip phases here
+let add5add2 = add 2 5
+(add5add2 10) = 17 // true
+
+// Or with more brevity
+(add 2 5 10) = 17 // true
+```
+
+As you can see, in ML-style languages, developer ergonomics are present and you can pass any number of arguments at once into a curried function. This style can be achieved in JavaScript using a helper function.
+
+```javascript
+const curry = (fn) =>
+  function _curry(...args) {
```

Comment:
You used function keyword again ;). See line 46.

Reply:
Yeah, had to create a named function here due to self referencing needs in the currying branch. Not an example though, since this is used in the actual examples. Gotta remove that aside though ;)